ttm/manager.py

```diff
@@ -0,0 +1,660 @@
+                'snapshotable': snapshotable,
+                'can_release': can_release,
+                'is_publishing': totest_is_publishing,
+            }
+            self.update_openqa_status_message()
+
+        if can_publish:
+            if current_qa_version == current_snapshot:
+                self.publish_factory_totest()
+                self.write_version_to_dashboard('snapshot', current_snapshot)
+                can_release = False  # we have to wait
+            else:
+                # We reached a very bad status: openQA testing is 'done', but not of the same version
+                # currently in test project. This can happen when 'releasing' the
+                # product failed
+                raise Exception('Publishing stopped: tested version (%s) does not match version in test project (%s)'
+                                % (current_qa_version, current_snapshot))
+
+        if can_release:
+            self.update_totest(new_snapshot)
+            self.write_version_to_dashboard('totest', new_snapshot)
+
+
+    def release(self, project):
+        self.setup(project)
+        new_snapshot = self.version_from_project()
+        self.update_totest(new_snapshot)
+
+    def write_version_to_dashboard(self, target, version):
+        version_file = 'version_%s' % target
+        if self.project.is_image_product:
+            version_file = version_file + '_images'
+        if not (self.dryrun or self.norelease):
+            self.api.pseudometa_file_ensure(version_file, version, comment='Update version')
+
+    def load_issues_to_ignore(self):
+        text = self.api.attribute_value_load('IgnoredIssues')
+        if text:
+            root = yaml.safe_load(text)
+            self.issues_to_ignore = root.get('last_seen')
+        else:
+            self.issues_to_ignore = dict()
+
+    def save_issues_to_ignore(self):
+        if self.dryrun:
+            return
+        text = yaml.dump({'last_seen': self.issues_to_ignore}, default_flow_style=False)
+        self.api.attribute_value_save('IgnoredIssues', text)
+
+    def release_version(self):
+        url = self.api.makeurl(['build', self.project.name, 'standard', self.project.arch,
+                                '000release-packages:%s-release' % self.project.base])
+        f = self.api.retried_GET(url)
+        root = ET.parse(f).getroot()
+        for binary in root.findall('binary'):
+            binary = binary.get('filename', '')
+            result = re.match(r'.*-([^-]*)-[^-]*.src.rpm', binary)
+            if result:
+                return result.group(1)
+
+        raise NotFoundException("can't find %s version" % self.project)
+
+    def version_from_project(self):
+        if not self.project.take_source_from_product:
+            return self.release_version()
+
+        if len(self.project.main_products):
+            return self.iso_build_version(self.project.name, self.project.main_products[0])
+
+        #assert(self.project.is_image_product)
```

Comment:
Should be removed

Reply:
it's not a leftover debug - it's a comment

Reply:
It's a commented out leftover debug statement then - this assert can't possibly be useful in production.

Reply:
Right, like most comments only help the reader to understand the logic - not the production

Reply:
You might be confused by this comment being pseudo code, but that's only because it's easier to express this way

Reply:
I reworked that whole image_product thingy now